### PR TITLE
Automatic font outlines (2)

### PR DESCRIPTION
**Pull request by ivan-mogilko**
This is a remake of @fernewelten's #892 PR. Old one still had several mistakes, but unfortunately I didn't know if and when fernewelten is planning to complete it, so decided to speed things up and fix it on my own.

Following is the original description:

---

_This is proposed code for issue #878 (automatic outlines)._
_Added two additional fields to fonts:_
- _`AutoOutlineStyle` (`Rounded`, `Squared`)._
- _`AutoOutlineThickness` (positive integer, or 0 for default)_ 
_(I couldn't use "1" as the default for `AutoOutlineThickness` because currently, scaled bitmap fonts get an outline that is thicker than 1)._ 

_The fields will only show in the editor if `OutlineStyle==Automatic`._
_The field `OutlineFont` will now only show if `OutlineStyle==UseOutlineFont`._

_The algo for generating the outline is based on messengerbag's. The text is only rendered _once_ per outline into a local bitmap; from then on, the bitmap is blitted repeatedly. Another local bitmap accumulates the vertical blits. In all, the number of blits is proportional to the thickness of the outline._ 

_For anti-aliased ttf fonts, a thickness of 2 to 3 seems to work nicely._

---

This PR consists of two parts: first are original fernewelten's commits (some minor later tweaks are squashed into bigger ones), and second part is my fixes following this list of issues: https://github.com/adventuregamestudio/ags/pull/892#issuecomment-522987666

**UPDATE**
Also, fixed following problem noted in fernewelten's original post:

> I couldn't use "1" as the default for `AutoOutlineThickness` because currently, scaled bitmap fonts get an outline that is thicker than 1

According to my research, this means the font outline is doubled (becomes = 2) under following condition:
* it is a hi-res game (according to legacy resolution rules);
* it is a bitmap font;
* it is a scaled up font (general settings have the deprecated "fonts are made for hi-res games" turned off, or newer format has individual scaling multiplier set to > 1).

It's possible to detect this condition during the format upgrade, in both editor and engine, so this will no longer be a concern.

### CODE REVIEW

**Comment by ivan-mogilko** [comment]
Note, I think I found a way to get rid of special value "0" for outline thickness and instead set thickness to either 1 or 2 depending on game and font settings during game import process (see UPDATE in the end of first post).

**Comment by ivan-mogilko** [comment]
How to test the old game import/run.

1. All fonts from old games should have outline = Squared.
2. Any fonts from low-res game (like 320x200) should have thickness = 1.
3. In hi-res game (like 640x400 and above) WFN fonts that either
a) made by prior to 3.5.0 and have "Fonts designed for hi-res games" setting = false, or
b) made in 3.5.0 and have SizeMultiplier > 1,
should have thickness = 2

**Comment by ivan-mogilko** [comment]
UPDATE: fixed following problem:

> New functionality (engine).
>* I found that if I set large thickness, then the engine cannot correctly calculate size of bitmap necessary to store the text line.
In my test this happened when I took font with height = 9 and assigned outline thickness = 9.

I believe it's important to note thought, that with larger outlines text lines are spread pretty far, which may be fixed by setting manual LineSpacing. When trying that I also found another potential future problem: ags does not know how "merge" outlines from different lines of text, and if they are too close the outline from the lower line overlap upper line. Guess the possible solution would demand a different drawing algorithm: where outline is drawn first for all the lines, and actual text is drawn only after (also for all the lines).

**Comment by ericoporto** [comment]
[test_outlines.zip](https://github.com/adventuregamestudio/ags/files/3998485/test_outlines.zip)

Screenshot from 2019-12-24 11-18-09

Hey, I did some tests, and things appear to work. On anti aliased fonts though, I get a weird artifact on the font at some specific points - on the white body at right you can see the font without outlines has no lines on top. Also on Roboto font, you can see some lines on outline on the back. Also dismiss the wrongly placed square there at the S letter, it's just a start from the background! xD

**Comment by ivan-mogilko** [comment]
> On anti aliased fonts though, I get a weird artifact on the font at some specific points - on the white body at right you can see the font without outlines has no lines on top.

Maybe it's alpha channel issue? and these vertical lines are where certain letter pixels became half-transparent and you can see black outline behind them? just a thought.
But did same happen before or it appeared only after these changes?

**Comment by ericoporto** [comment]
Screenshot from 2019-12-24 14-01-28

Above is the previous result - they didn't had artifacts. Click on the images to load on a new tab and expand them.

Some comments on the c++ code, bear in mind I had difficulties trying to follow it and understand why it's as is:
- wouttextxy_AutoOutline arguments are reversed when comparing to other functions.
- `Bitmap *outline_stencil = CreateTransparentBitmap(t_width, t_height + 2 * thickness, ds->GetColorDepth());` : Why it doesn't have the same 2 times thickness on the width too?
-   why this below is done in two lines
```
wouttextxy(texx_stencil, 0, 0, font, color, texx);
wouttextxy_AutoOutline_Semitransparent2Opaque(texx_stencil);
```
- shouldn't the blit to ds only happen once?

**Comment by ivan-mogilko** [comment]
Oh, I did not look into the drawing itself yet, I will check that out.

**Comment by ericoporto** [comment]
Uhm, other thing, because on AGS we can't dynamically change font properties like font size, or in this case the thickness of the outline, if you feel you need to, it's alright to pre-generate the outline on the Editor - at build time, if it's not updated.

**Comment by ivan-mogilko** [comment]
> Uhm, other thing, because on AGS we can't dynamically change font properties like font size, or in this case the thickness of the outline, if you feel you need to, it's alright to pre-generate the outline on the Editor - at build time, if it's not updated.

I don't think this kind of optimization should be done in the current PR.
Also, doing this on Editor's side would mean extra compilation step and creating a font cache, with all the consequences of maintaining it, so it's a serious change (and Idk how to do that for TTF fonts).

**Comment by ericoporto** [comment]
image

No worries. Just to emphasize the artifacts, above are weird horizontal lines in the outline itself, and on the text at front, the vertical lines.

**Comment by ivan-mogilko** [comment]
@fernewelten , if you see this, could you explain the purpose of using wouttextxy_AutoOutline_Semitransparent2Opaque()?

**Comment by ericoporto** [comment]
What does this blit thing does? I feel the bits are being laid on top of other without being actually blended.

**Comment by ivan-mogilko** [comment]
> What does this blit thing does? I feel the bits are being laid on top of other without being actually blended.

I am pretty much sure this is the case. For blending there are other functions.

**Comment by ericoporto** [comment]
OK, this means the resulting alpha is always the last one thrown at that pixel.

**Comment by ivan-mogilko** [comment]
@ericoporto 

> shouldn't the blit to ds only happen once?

if you are referring to this:
<pre> // stamp the outline stencil to the left and right of the text
        ds->Blit(outline_stencil, xxp - x_diff, outline_y, kBitmap_Transparency);
        if (x_diff > 0)
            ds->Blit(outline_stencil, xxp + x_diff, outline_y, kBitmap_Transparency);</pre>
This code draws dark font with left and right shift so that outline appears on both sides horizontally.

**Comment by ivan-mogilko** [comment]
Alright, I was probably confused at first. What `Semitransparent2Opaque` is used for - is to increase opaqueness of the _outline shadow_. How may come it is translucent in the first place? Is it what may happen if you outline piece with a TTF font (or anti-aliased TTF font)?

But in this case, should not it be simply made 100% opaque by stripping alpha channel without this extra algorithm?

**Comment by ericoporto** [comment]
I found something online -> https://d1cxvcw9gjxu2x.cloudfront.net/attachments/606810

It was here: ttpss://www.allegro.cc/forums/thread/611384

Not sure if relevant or if it works.

**Comment by ivan-mogilko** [comment]
I begin to doubt that this PR should be adding a special algorithm for TTF outlines at all. This seem to be superflous for now, and maybe better left until real problem is stated. Perhaps it will be found more optimal to actually use outlining abilities of TTF library.

**Comment by ivan-mogilko** [comment]
NOTE: I adjusted your test game by reducing resolution and adding another non-transparent GUI underneath the labels to make errors more visible:
[test_outlines_640x480.zip](https://github.com/adventuregamestudio/ags/files/3999118/test_outlines_640x480.zip)

**Comment by messengerbag** [comment]
> Alright, I was probably confused at first. What `Semitransparent2Opaque` is used for - is to increase opaqueness of the _outline shadow_. How may come it is translucent in the first place? Is it what may happen if you draw outline piece with a TTF font (or anti-aliased TTF font)?

Yes, exactly. The naive way of drawing outlines for anti-aliased fonts tends to create semi-opaque sections (I've experienced this with the SpeechBubble module). The function looks like an attempt to turn the outline version opaque -- compare this comment:

https://github.com/adventuregamestudio/ags/issues/878#issuecomment-515603670

I would also assume there is a better way. (I would also add that for the best results, there should be two versions of the outline stamp: the fully opaque one used to fill the outline, and the anti-aliased one used along the edge of the outline so it will also be anti-aliased.)

But if there is a way to do this outlining in the font library I would think that would be better.

**Comment by ericoporto** [comment]
A second step after this PR is done is probably text caching, so if you draw something, next time you just get it from the cache if it's there draw it instead of constructing it again. I don't remember how text drawing works in AGS, if it's figured on every frame or only on the first call.

**Comment by ivan-mogilko** [comment]
> I don't remember how text drawing works in AGS, if it's figured on every frame or only on the first call.

Can't tell for 100% cases, but in majority of them the text and other elements are only prepared once, and then sent into renderer each frame while the message is same and image does not have to be updated for any reason.

**Comment by ericoporto** [comment]
Gave a try, but it didn't work :grimacing: 

```C++
// Draw outline that is calculated from the text font, not derived from an outline font
void wouttextxy_AutoOutline(Bitmap *ds, size_t font, int32_t color, const char *texx, int &xxp, int &yyp)
{
    using AGS::Common::kBitmap_Transparency;

    int thickness = game.fonts.at(font).AutoOutlineThickness;
    auto style = game.fonts.at(font).AutoOutlineStyle;
    if (thickness <= 0)
        return; 

    size_t const t_width = wgettextwidth(texx, font);
    size_t const t_height = wgettextheight(texx, font);

    
    // move start of text so that the outline doesn't drop off the bitmap
    xxp += thickness;
    int const outline_y = yyp;
    yyp += thickness;
    
    int largest_y_diff_reached_so_far = -1; 
    for (int x_diff = thickness; x_diff >= 0; x_diff--)
    {
        // Integer arithmetics: In the following, we use terms k*(k + 1) to account for rounding.
        //     (k + 0.5)^2 == k*k + 2*k*0.5 + 0.5^2 == k*k + k + 0.25 ==approx. k*(k + 1)
        int y_term_limit = thickness * (thickness + 1);
        if (FontInfo::kRounded == style)
            y_term_limit -= x_diff * x_diff;

        // extend the outline stencil to the top and bottom
        for (int y_diff = largest_y_diff_reached_so_far + 1; 
            y_diff <= thickness && y_diff * y_diff  <= y_term_limit; 
            y_diff++)
        {
            printf("x_diff=%d , y_diff=%d, text=%s \n", x_diff,y_diff, texx);

            wouttextxy(ds, xxp - x_diff, outline_y + thickness - y_diff, font, color, texx);
            wouttextxy(ds, xxp, outline_y + thickness - y_diff, font, color, texx);
            if (x_diff > 0)
                wouttextxy(ds, xxp + x_diff, outline_y + thickness - y_diff, font, color, texx);


            wouttextxy(ds, xxp - x_diff, outline_y + thickness , font, color, texx);
            wouttextxy(ds, xxp, outline_y + thickness , font, color, texx);
            if (x_diff > 0)
                wouttextxy(ds, xxp + x_diff, outline_y + thickness , font, color, texx);

            if (y_diff > 0) {
                wouttextxy(ds, xxp - x_diff, outline_y + thickness + y_diff, font, color, texx);
                wouttextxy(ds, xxp, outline_y+ thickness + y_diff, font, color, texx);
                if (x_diff > 0)
                    wouttextxy(ds, xxp + x_diff, outline_y + thickness + y_diff, font, color, texx);
            }
            largest_y_diff_reached_so_far = y_diff;
        }
    }
}
```

Console log
```bash
x_diff=5 , y_diff=0, text=test_outlines 
x_diff=5 , y_diff=1, text=test_outlines 
x_diff=5 , y_diff=2, text=test_outlines 
x_diff=4 , y_diff=3, text=test_outlines 
x_diff=3 , y_diff=4, text=test_outlines 
x_diff=2 , y_diff=5, text=test_outlines 
x_diff=5 , y_diff=0, text=Score: 0 
x_diff=5 , y_diff=1, text=Score: 0 
x_diff=5 , y_diff=2, text=Score: 0 
x_diff=4 , y_diff=3, text=Score: 0 
x_diff=3 , y_diff=4, text=Score: 0 
x_diff=2 , y_diff=5, text=Score: 0 
x_diff=6 , y_diff=0, text=Test Roboto Font 
x_diff=6 , y_diff=1, text=Test Roboto Font 
x_diff=6 , y_diff=2, text=Test Roboto Font 
x_diff=6 , y_diff=3, text=Test Roboto Font 
x_diff=6 , y_diff=4, text=Test Roboto Font 
x_diff=6 , y_diff=5, text=Test Roboto Font 
x_diff=6 , y_diff=6, text=Test Roboto Font 
x_diff=4 , y_diff=0, text=Girassol Font Test 
x_diff=4 , y_diff=1, text=Girassol Font Test 
x_diff=4 , y_diff=2, text=Girassol Font Test 
x_diff=3 , y_diff=3, text=Girassol Font Test 
x_diff=2 , y_diff=4, text=Girassol Font Test 
```

image

**Comment by messengerbag** [comment]
Well, that picture at least shows pretty clearly what the problem is (or one of them, anyway): The horizontal displacement only draws the leftmost, rightmost and center (undisplaced) copies, rather than looping all the way from left to right.

**Comment by ericoporto** [comment]
I tried removing the "ifs" and it didn't work too... Found a different algorithm, but think it's not applicable here: https://stackoverflow.com/questions/1109536/an-algorithm-for-inflating-deflating-offsetting-buffering-polygons

**Comment by ivan-mogilko** [comment]
@ericoporto what are you trying or suggesting to do exactly?

I am planning to do speed test with `wouttextxy` vs current implementation or current implementation with Blits replaced with proper blending (did not yet test that it actually works). But I have not had time today and will probably try tomorrow.

**Comment by ivan-mogilko** [comment]
@ericoporto only for the reference, here's the proper blending function I mentioned earlier. It's definitely going to be somewhat slower than the Blit one though.

Only changes compared to original are:
* included "gfx/blender.h"
* `set_blender_mode(nullptr, nullptr, _argb2argb_blender, 0, 0, 0, 0xff);` call before the drawing
* replaced all Blit with TransBlend.

<details><summary>CLICK ME</summary>
<p>

```C++
#include "gfx/blender.h"

void wouttextxy_AutoOutline_alphablend(Bitmap *ds, size_t font, int32_t color, const char *texx, int &xxp, int &yyp)
{
    int thickness = game.fonts.at(font).AutoOutlineThickness;
    auto style = game.fonts.at(font).AutoOutlineStyle;
    if (thickness <= 0)
        return;

    size_t const t_width = wgettextwidth(texx, font);
    size_t const t_height = wgettextheight(texx, font);
    Bitmap *outline_stencil =
        CreateTransparentBitmap(t_width, t_height + 2 * thickness, ds->GetColorDepth());
    Bitmap *texx_stencil =
        CreateTransparentBitmap(t_width, t_height, ds->GetColorDepth());
    if (!outline_stencil || !texx_stencil)
        return;
    wouttextxy(texx_stencil, 0, 0, font, color, texx);

    // move start of text so that the outline doesn't drop off the bitmap
    xxp += thickness;
    int const outline_y = yyp;
    yyp += thickness;

    set_blender_mode(nullptr, nullptr, _argb2argb_blender, 0, 0, 0, 0xff);

    int largest_y_diff_reached_so_far = -1;
    for (int x_diff = thickness; x_diff >= 0; x_diff--)
    {
        // Integer arithmetics: In the following, we use terms k*(k + 1) to account for rounding.
        //     (k + 0.5)^2 == k*k + 2*k*0.5 + 0.5^2 == k*k + k + 0.25 ==approx. k*(k + 1)
        int y_term_limit = thickness * (thickness + 1);
        if (FontInfo::kRounded == style)
            y_term_limit -= x_diff * x_diff;

        // extend the outline stencil to the top and bottom
        for (int y_diff = largest_y_diff_reached_so_far + 1;
            y_diff <= thickness && y_diff * y_diff <= y_term_limit;
            y_diff++)
        {
            outline_stencil->TransBlendBlt(texx_stencil, 0, thickness - y_diff);
            if (y_diff > 0)
                outline_stencil->TransBlendBlt(texx_stencil, 0, thickness + y_diff);
            largest_y_diff_reached_so_far = y_diff;
        }

        // stamp the outline stencil to the left and right of the text
        ds->TransBlendBlt(outline_stencil, xxp - x_diff, outline_y);
        if (x_diff > 0)
            ds->TransBlendBlt(outline_stencil, xxp + x_diff, outline_y);
    }

    delete texx_stencil;
    delete outline_stencil;
}
```
</p>
</details>

**Comment by ericoporto** [comment]
Hey, xml version 26 is kept for some patching on 3.5.0.23+ ?

**Comment by ivan-mogilko** [comment]
> Hey, xml version 26 is kept for some patching on 3.5.0.23+ ?

Yes, 26 is already taken by 3.5.0 patch. I will rebase this PR when patch is released and merged to master.

Again for the reference, here's probably the correct variant that only uses direct text output:

<details><summary>CLICK ME</summary>
<p>

```C++

void wouttextxy_AutoOutline_dumb(Bitmap *ds, size_t font, int32_t color, const char *texx, int &xxp, int &yyp)
{
    int thickness = game.fonts.at(font).AutoOutlineThickness;
    auto style = game.fonts.at(font).AutoOutlineStyle;
    if (thickness <= 0)
        return;

    size_t const t_width = wgettextwidth(texx, font);
    size_t const t_height = wgettextheight(texx, font);

    // move start of text so that the outline doesn't drop off the bitmap
    xxp += thickness;
    yyp += thickness;

    for (int x_diff = thickness; x_diff >= 0; x_diff--)
    {
        // Integer arithmetics: In the following, we use terms k*(k + 1) to account for rounding.
        //     (k + 0.5)^2 == k*k + 2*k*0.5 + 0.5^2 == k*k + k + 0.25 ==approx. k*(k + 1)
        int y_term_limit = thickness * (thickness + 1) + 1;
        if (FontInfo::kRounded == style)
            y_term_limit -= x_diff * x_diff;

        for (int y_diff = 0; y_diff <= thickness
            && y_diff * y_diff <= y_term_limit;
            y_diff++)
        {
            wouttextxy(ds, xxp - x_diff, yyp - y_diff, font, color, texx);
            wouttextxy(ds, xxp - x_diff, yyp + y_diff, font, color, texx);
            if (x_diff > 0)
            {
                wouttextxy(ds, xxp + x_diff, yyp - y_diff, font, color, texx);
                if (y_diff > 0)
                    wouttextxy(ds, xxp + x_diff, yyp + y_diff, font, color, texx);
            }
        }
    }
}
```
</p>
</details>

Compared to the Blit or TransBlend versions this gives more jagged edges, my guess that is because of how alpha composition works inside alfont renderer.

IMHO for visual test it's best to arrange pairs of lines with same font but different outline types.

**Comment by ivan-mogilko** [comment]
I did  some speed tests in a separate branch: https://github.com/ivan-mogilko/ags-refactoring/commit/1616d4a0effc063151fc61604f4cf5b7ecdfb75b

Was drawing same text line with an outline 1000 times each, trying 5 different functions:
- stencils - original fernewelten's function using intermediate stencil bitmaps and simple blits (not suitable for anti-aliased TTFs)
- stencils alphablend - variant of above meant specially for antia-aliased TTFs in 32-bit games.
- singlestencil - variant of above but using only 1 stencil with text on it (does more drawing on destination bitmap, but not creates another intermediate bitmap).
- singlestencil alphablend - "singlestencil" but for AA TTFs.
- wouttextxy only (dumb) - no intermediate bitmaps, font renderer fills all the outline to the destination (works rather well for all fonts, but with TTFs it may produce jagged edges).

WFN (bitmap) font (roughly same height), auto outline thickness = 4

* stencils (1000 times): 979 (ms)
* stencils alphablend (1000 times): 937 (ms)
* singlestencil (1000 times): 2049 (ms)
* singlestencil alphablend (1000 times): 25050 (ms) (!!!)
* wouttextxy only (dumb) (1000 times): 3608 (ms)

TTF font, NO anti-aliasing, auto outline thickness = 4

* stencils (1000 times): 1215 (ms)
* stencils alphablend (1000 times): 1094 (ms)
* singlestencil (1000 times): 2180 (ms)
* singlestencil alphablend (1000 times): 32718 (ms)
* wouttextxy only (dumb) (1000 times): 9776 (ms)

Same TTF font, with anti-aliasing, auto outline thickness = 4

* stencils (1000 times): 1382 (ms)
* stencils alphablend (1000 times): 3390 (ms)
* singlestencil (1000 times): 2361 (ms)
* singlestencil alphablend (1000 times): 49916 (ms)
* wouttextxy only (dumb) (1000 times): 21785 (ms)

**Comment by AlanDrake** [comment]
> One amazing thing from above is that alpha-blending is actually slightly faster (?!) 
Could it be compiler optimizations?
I remember my old software filters were much faster with bad code rathen than the version where I had tried to optimize, in the end  it was only thanks to compiler optimizations.

**Comment by messengerbag** [comment]
Hmm, the measurements for the "singlestencil" method seem to have disappeared, but I'm a bit surprised by the values given initially. It seems to take a little more than twice as long, while by my estimate it should perform (4+1+4)^2=81 draw operations to create the outline, vs. only 2*(4+1+4)=18 for the "stencils" method: 4.5 times as many. I suppose writing to the temp bitmap might add some overhead, but not that much, surely?

**Comment by ivan-mogilko** [comment]
Yes, it was actually weird, and I was suspecting doing something wrong there, but unfortunately was out of spare time to look again. Idk if it's worth now though, because original method seem to outperform everything else.

My only concern was (an is) that it continiously allocates temp bitmaps. Perhaps later we could consider a cache for these, assuming that many texts on screen may stay unchanged for a while and therefore have same sizes (such as menu texts, static gui labels etc).

**Comment by ivan-mogilko** [comment]
@messengerbag Ah, thanks for noticing, I found that I used a simple Blit there instead of Blit with transparency.

New results look like following

<details><summary>CLICK ME</summary>

WFN

* stencils (1000 times): 882 (ms)
* stencils alphablend (1000 times): 911 (ms)
* singlestencil (1000 times): 15679 (ms)
* singlestencil alphablend (1000 times): 24351 (ms)
* wouttextxy only (dumb) (1000 times): 3602 (ms)

TTF font, NO anti-aliasing

* stencils (1000 times): 1181 (ms)
* stencils alphablend (1000 times): 1134 (ms)
* singlestencil (1000 times): 21261 (ms)
* singlestencil alphablend (1000 times): 31048 (ms)
* wouttextxy only (dumb) (1000 times): 9804 (ms)

Same TTF font, with anti-aliasing, auto outline thickness = 4

* stencils (1000 times): 1319 (ms)
* stencils alphablend (1000 times): 3385 (ms)
* singlestencil (1000 times): 22194 (ms)
* singlestencil alphablend (1000 times): 50117 (ms)
* wouttextxy only (dumb) (1000 times): 22157 (ms)

</details>

**Comment by ivan-mogilko** [comment]
@messengerbag actually, that "singlestencil" function was stupid, I made an alternate one that removes different bitmap (https://github.com/ivan-mogilko/ags-refactoring/commit/7d5a052c51aac81d49c1cd4502d47d269d9d73d2). According to tests it's much closer to best results, but still not as fast:

<details><summary>CLICK ME</summary>

WFN

* singlestencil2 (1000 times): 900 (ms)
* singlestencil2 alphablend (1000 times): 1295 (ms)

TTF font, NO anti-aliasing

* singlestencil2 (1000 times): 1700 (ms)
* singlestencil2 alphablend (1000 times): 2169 (ms)

Same TTF font, with anti-aliasing

* singlestencil2 (1000 times): 3099 (ms)
* singlestencil2 alphablend (1000 times): 3873 (ms)

</details>

**Comment by ericoporto** [comment]
Hey, I think the results for the Engine are good, could select one that is visually correct and is reasonably fast and commit. 

On the Editor I noticed only minor things, I will comment, but they may be dismissed.
- some inconsistency with identation on the file `Editor/AGS.Types/Font.cs`. 
- I think the content of the  `PropertyDescriptorCollection GetProperties(Attribute[] attributes)` call is there to hide properties depending of the condition? Maybe a comment there just to clarify this intent.

Actual functionality appears correct for me.

**Comment by ivan-mogilko** [comment]
For 32-bit it seems the "alphablend" variant makes good results, both in terms of speed and looks. Speed is close to blit version in average for non-AA fonts, and it's the variant that draws correctly for AA-fonts, so it should be okay to use for all of them. AA outline is also created similar, if not identical, to the one drawn by rendering a font directly. So all good here.

For 16-bit variant atm I was not able to achieve "proper" looks for AA TTF fonts, because in 16-bit you cannot have alpha channel, and text has to be drawn directly on final surface to blend with the background colors, but that is not happening if we are using a transparent bitmap ("stencil") in between. So the text itself is anti-aliased, but outline is not.

Did not test 8-bit version yet. Anti-aliasing is disabled for 8-bit games, so only need to check non-AA font rendering.

**Comment by ivan-mogilko** [comment]
> For 32-bit it seems the "alphablend" variant makes good results, both in terms of speed and looks. Speed is close to blit version in average for non-AA fonts

So, when trying to make new outlines work in 16-bit, I found that I did a stupid mistake earlier, and was passing wrong "color" value without alpha component. Upon fixing, "alphablend" function became noticeably slower.
I was planning on writing a universal function that switches between blending and blitting anyway because of 16- and 8-bit games, which I did (latest commits here: https://github.com/ivan-mogilko/ags-refactoring/commits/ags3--AFO-tests).
Now the blending is only done for anti-aliased TTF fonts (in 16->32 bit games), and masked blit is done in all other cases.

Latest speed test results:
[test outlines 8bit -- with universal funs.txt](https://github.com/adventuregamestudio/ags/files/4014187/test.outlines.8bit.--.with.universal.funs.txt)
[test outlines 16bit -- with universal funs.txt](https://github.com/adventuregamestudio/ags/files/4014188/test.outlines.16bit.--.with.universal.funs.txt)
[test outlines 32bit -- with universal funs.txt](https://github.com/adventuregamestudio/ags/files/4014189/test.outlines.32bit.--.with.universal.funs.txt)

Notes:
- single stencil variant is actually faster with WFN fonts, but not much faster and the relative difference decreases with the thickness.

**Comment by ivan-mogilko** [comment]
Pushed universal function that should support anti-aliased outlines in 32-bit and 16-bit games, but also falls back to regular blits if font is not anti-aliased.

File: Editor/AGS.Types/Font.cs [hunk 1]
```diff
@@ -120,7 +120,7 @@ public FontOutlineStyle OutlineStyle
             set { _outlineStyle = value; }
         }
 
-        [Description("Thickness of the automatic outline")]
+        [Description("Thickness of the automatic outline. WARNING: automatic outlines are generated at runtime and large sizes may negatively affect your game perfomance.")]
```
Comment (morganwillcock) [inline Editor/AGS.Types/Font.cs]: There is a typo here as 'performance' is missing the 'r'.

**Pull request merged**